submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,359 @@
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type ClusterGlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of ClusterGlobalEgressIP object.
+    Spec               ClusterGlobalEgressIPSpec `json:"spec"`
+
+    // Observed status of ClusterGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type ClusterGlobalEgressIPSpec struct {
+    // The number of globalIP's requested at the cluster-level.
+    // Globalnet Controller will allocate the requested number of contiguous GlobalIPs for
+    // this ClusterGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
+    // +optional
+    NumGlobalIPs  *int     'json:"numGlobalIPs",omitempty"`
+}
+
+type GlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalEgressIP object.
+    Spec               GlobalEgressIPSpec `json:"spec"`
+
+    // Observed status of GlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type GlobalEgressIPSpec struct {
+    // The number of globalIP's requested at the namespace level or for selected pods in a namespace.
+    // Globalnet Controller will allocate the requested number of contiguous GlobalIPs for this
+    // GlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
+    // +optional
+    NumGlobalIPs  *int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the namespace in which the
+    // CR is created.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+const (
+    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
+    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"
+
+    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
+    GlobalEgressIPError GlobalEgressIPStatus = "Error"
+)
+
+type EgressIPStatus struct {
+    // Status is one of {"Success", "Error"}
+    Status GlobalEgressIPStatus `json:"status,omitempty"`
+
+    // +optional
+    Message *string `json:"message,omitempty"`
+
+    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
+    GlobalIPs []string `json:"globalIPs"`
```

Comment:
How about `AllocatedIPs`?

Reply:
At lot of places in this PR as well as in our doc we use globalIP or globalIPs. But if others prefer `AllocatedIPs`, I can change it.

Reply:
We can still use the term _globalIP_ - `AllocatedIPs` would refer to the globalIPs allocated for this instance. I only mentioned this b/c we decided to drop _Global_ from `NumIPs` so perhaps we should do so here for consistency.

Reply:
Its fine to me. Lets hear from @mangelajo @mkimuram

Reply:
As for the consistency, `GlobalEgressIP` now contains `GlobalEgressIPSpec` and `EgressIPStatus`, it sounds a bit inconsistent. 

I don't have a strong preference on removing `Global` from everywhere, but it would be possible to just 's/Global//', except `GlobalIngressIPStatus.GlobalIP` (it should be something like `IngressIPStatus.AllocatedIP`), without conflict and significant meaning change (unless there are non-global EgressIP or IngressIP object for other purpose).

The other direction is just move `EgressIPStatus` back to `EgressGlobalIPStatus`, but I prefer above way. In both case, I agree with changing to `AllocatedIP`(s).

Reply:
In https://github.com/submariner-io/submariner/pull/1172, I named it `GlobalEgressIPStatus` for consistency. I thought about dropping the `Global` prefix from the structs but doc comments refer to "_GlobalIPs_" and "_Globalnet_" so it seems they're implicitly tied to Globalnet.

Reply:
Thank you @mkimuram @tpantelis. I'll use AllocatedIPs

Reply:
Ok I pushed https://github.com/submariner-io/submariner/pull/1334 to update the types.  I'd like to suggest that we continue reviewing the type implementations in that PR. I think this PR should be primarily for concepts plus this one has gotten quite large 😄